### PR TITLE
refactor(worker, dal): implement atomic job claiming and child job queuing fixes NV-7597

### PR DESCRIPTION
**Pull request by merrcury**
## What changed

This PR implements atomic job claiming to prevent duplicate job processing and child job enqueueing under concurrent worker execution. Two new atomic claim methods were added to `JobRepository`: `claimAsRunning()` transitions jobs from QUEUED/DELAYED to RUNNING, and `claimNextChildAsQueued()` transitions child jobs from PENDING to QUEUED. The `RunJob` use case was refactored to use atomic claims instead of separate read-then-update operations, ensuring only one worker can successfully claim a job.

## Affected areas

**`dal`**: Added two atomic claim methods to `JobRepository` that use `findOneAndUpdate` to atomically transition job statuses, returning the claimed job or `null` if no matching job is found.

**`worker`**: Updated `RunJob.execute()` to atomically claim the job as `RUNNING` at the start, exiting early if another worker already claimed it. Refactored `tryQueueNextJobs()` to use atomic child job claiming instead of a separate lookup to prevent duplicate child enqueueing from redelivered messages.

## Key technical decisions

- Used MongoDB `findOneAndUpdate` operations for atomic state transitions rather than separate fetch-then-update patterns, eliminating race conditions between concurrent workers.
- Changed child job selection from `jobRepository.findOne()` to `claimNextChildAsQueued()` to prevent duplicate enqueueing on retries/redeliveries.
- Simplified execution flow by removing explicit status update calls after atomic claims, since the claimed job is already in the target state.

## Testing

Added comprehensive tests for atomic claiming behavior including: transitions between job statuses (QUEUED → RUNNING, DELAYED → RUNNING, PENDING child → QUEUED), null returns when jobs are not claimable, and exclusivity under concurrent callers (only one claim succeeds). Tests were also updated to set job status to QUEUED directly, bypassing the normal `AddJobUsecase` flow.

[Review Change Stack](https://app.coderabbit.ai/change-stack/novuhq/novu/pull/11094)

### Screenshots

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Special notes for your reviewer

</details>

<h3>Greptile Summary</h3>

This PR moves standard worker job execution to atomic MongoDB claims. The main changes are:

- Atomic `RUNNING` claims for queued, delayed, and stale running jobs.
- Heartbeat renewal and release helpers for running job leases.
- Atomic claiming for pending child jobs before enqueueing the next step.
- Recovery for redelivered completed parent jobs with still-pending children.
- Tests for claim exclusivity, stale lease recovery, heartbeat cleanup, and redelivery recovery.

<h3>Confidence Score: 5/5</h3>

Safe to merge with low risk.

The changed paths address the reviewed atomic-claim and redelivery edge cases with lease renewal, stale-claim recovery, child-claim serialization, and focused tests. No verified functional or security issues remain in the reviewed changes.

No files require special attention.

<details><summary><h3><a href="https://www.greptile.com/trex">T-Rex</a> T-Rex Logs</h3></summary>

**What T-Rex did**
- Observed that the narrow worker spec command ran and was killed with SIGKILL before any assertion output.
- Captured blocker evidence showing that mongod, redis-server, and docker are not installed or not found in PATH, which would prevent service checks from running.
- Captured partial proof that the @novu/dal build completes successfully with exit code 0.

<a href="https://app.greptile.com/trex/runs/15782357/artifacts"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifactsDark.svg?v=4"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifacts.svg?v=4">View all artifacts</picture></a>

<sub><a href="https://www.greptile.com/trex">T-Rex</a> Ran code and verified through T-Rex</sub>
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/worker/src/app/workflow/usecases/run-job/run-job.usecase.ts | Refactors job execution to use atomic running claims, heartbeat lease renewal, completed-parent recovery, and atomic child claiming. |
| libs/dal/src/repositories/job/job.repository.ts | Adds atomic job claim, lease renewal, lease release, child claim helpers, and includes queued digest jobs in backoff lookups. |
| apps/worker/src/app/workflow/services/standard.worker.spec.ts | Updates worker tests for queued job setup and adds coverage for claims, lease renewal, heartbeat cleanup, and redelivery recovery. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
participant Queue as Standard Queue
participant Worker as RunJob
participant Jobs as JobRepository
participant AddJob as AddJob
participant DB as MongoDB

Queue->>Worker: deliver job id
Worker->>Jobs: findOne(job)
Worker->>Jobs: claimAsRunning(environmentId, jobId)
Jobs->>DB: "findOneAndUpdate QUEUED/DELAYED/stale RUNNING -> RUNNING"
alt claim succeeds
    Worker->>Jobs: renewRunningClaim heartbeat while executing
    Worker->>Worker: execute step and persist status
    Worker->>Jobs: claimNextChildAsQueued(parentId)
    Jobs->>DB: "findOneAndUpdate PENDING child -> QUEUED"
    alt child claimed
        Worker->>AddJob: enqueue claimed child
        AddJob->>Queue: queue child job
    else no child
        Worker->>Worker: finalize workflow run
    end
else claim fails
    Worker->>Jobs: check completed parent for PENDING child
    Worker->>Jobs: claimNextChildAsQueued(parentId)
    Worker->>AddJob: resume stranded child if present
end
```

<sub>Reviews (5): Last reviewed commit: ["refactor(worker, tests): enhance RunJob ..."](https://github.com/novuhq/novu/commit/3db131561e7969b4620266f213b862e6828a24f9) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=33899642)</sub>

### CODE REVIEW

**Comment by merrcury** [comment]
Manual end-to-end redelivery test against a real worker + real Mongo

Trigger a 3-step in-app workflow via POST /v1/events/trigger.

- Poll Mongo jobs collection until all 4 jobs (1 trigger + 3 channel) reach a terminal state.
- Snapshot messages count grouped by _jobId → expected 3.
- Pick one already-completed channel job and re-enqueue it onto the standard queue twice with the exact data shape extractMinimalJobData expects (_environmentId, _id, _organizationId, _userId).
- Sleep 5s for the worker to pick up both redelivery messages.
- Snapshot again — assert messages count is unchanged and the redelivered job's status is unchanged.
- Observed result of the manual test

Before redelivery: 4 jobs, 3 messages.

- After 2× redelivery of one completed in-app job: 4 jobs, 3 messages (no duplicates).
- Worker logs the new code path twice (once per redelivery): Skipping job: could not atomically claim (already running, completed, or canceled by another worker).

What this validates

- Application-level at-most-once execution under simulated SQS at-least-once redelivery
- The atomic claim is exclusive under real concurrency, not just in unit tests.
- No duplicate sendMessage, no duplicate MessageEntity, no resurrection of completed jobs.
- claimNextChildAsQueued prevents re-enqueueing an already-claimed child step in the parent-redelivery scenario.


Screenshot 2026-05-12 at 20 04 57

**Review by scopsy — APPROVED** [review]

File: apps/worker/src/app/workflow/usecases/run-job/run-job.usecase.ts [hunk 1]
```diff
@@ -480,10 +498,14 @@ export class RunJob {
...
-          _parentId: currentJob._id,
-        });
+        /*
+         * Atomic child claim: transition the child PENDING -> QUEUED in one
+         * doc op. A redelivered parent that re-enters this loop cannot find a
+         * still-PENDING child (the first run already claimed it), so it cannot
+         * re-enqueue the same logical step. addJobUsecase.execute() below will
+         * then re-set the status to QUEUED/DELAYED idempotently.
+         */
+        nextJob = await this.jobRepository.claimNextChildAsQueued(currentJob._environmentId, currentJob._id);
```
[the remark this answers is not in the document]
Comment (merrcury) [reply apps/worker/src/app/workflow/usecases/run-job/run-job.usecase.ts]: @coderabbitai Existing catch + `setJobAsFailed` covers the common path; remaining gap is a reaper concern, not a per-call invariant.

[the remark this answers is not in the document]
Comment (merrcury) [reply apps/worker/src/app/workflow/usecases/run-job/run-job.usecase.ts]: @coderabbitai resolve this

File: libs/dal/src/repositories/job/job.repository.ts [hunk 2]
```diff
@@ -74,6 +74,59 @@ export class JobRepository extends BaseRepository<JobDBModel, JobEntity, Enforce
...
+        _parentId: parentJobId,
+        status: JobStatusEnum.PENDING,
+      },
+      {
+        $set: {
+          status: JobStatusEnum.QUEUED,
+        },
+      },
+      { new: true }
+    );
```
[the remark this answers is not in the document]
Comment (merrcury) [reply libs/dal/src/repositories/job/job.repository.ts:161]: @coderabbitai 

- Fan-out is structurally tiny. The _parentId cardinality is "one parent → typically 1 child" (a workflow step has one next step). Even in skip-cascade workflows, you're scanning at most 3-5 candidates from the _parentId index. That's not a "collection scan", it's an index seek + sub-millisecond filter on a few documents.
- Adding the compound index has real write cost. Every job insert/update writes to one more index. Standard queue ingestion is high-volume; this matters.
- The repo's index philosophy (see all the comments in job.schema.ts) is "add an index when a query path is shown to be slow," not "add an index for every multi-field query."


Let me know if you think otherwise

[the remark this answers is not in the document]
Comment (merrcury) [reply libs/dal/src/repositories/job/job.repository.ts:161]: @coderabbitai resolve this

File: apps/worker/src/app/workflow/usecases/run-job/run-job.usecase.ts [hunk 3]
```diff
@@ -540,10 +558,14 @@ export class RunJob {
...
```
[the remark this answers is not in the document]
Comment (merrcury) [reply apps/worker/src/app/workflow/usecases/run-job/run-job.usecase.ts:615]: Partially addressed in e2366b8986: a redelivered COMPLETED parent now resumes the chain via `resumeChainIfStrandedAfterCompletion` (claims the still-PENDING child idempotently). The residual window — crash between the child claim and the actual enqueue — mirrors the pre-existing window inside `AddJob` itself (status is set to QUEUED/DELAYED before `queueJob`), and re-running AddJob on a QUEUED child would duplicate digest-merge evaluation, so it needs an idempotent-enqueue design rather than a wider claim predicate. Tracked in NV-8394.

**Comment by merrcury** [comment]
Re the secondary note about `claimNextChildAsQueued` handing AddJob a pre-mutated QUEUED job: traced it — nothing in `AddJob` (or the SKIPPED continuation in `tryQueueNextJobs`) branches on the incoming `status === PENDING`. The only `job.status` reads in the worker usecases are RunJob's CANCELED check and the digest usecase filtering COMPLETED/FAILED next jobs, so the shift from PENDING to QUEUED is behaviorally inert.

File: apps/worker/src/app/workflow/usecases/run-job/run-job.usecase.ts [hunk 4]
```diff
@@ -125,6 +122,36 @@ export class RunJob {
...
+    const claimHeartbeat = setInterval(() => {
+      this.jobRepository.renewRunningClaim(claimed._environmentId, claimed._id).catch((renewError: unknown) => {
+        this.logger.warn({ err: renewError, nv: { jobId: claimed._id } }, 'Failed to renew running claim lease');
+      });
+    }, RUNNING_CLAIM_RENEW_INTERVAL_MS);
+    claimHeartbeat.unref();
+
+    await this.stepRunRepository.create(job, {
+      status: JobStatusEnum.RUNNING,
+    });
```
[the remark this answers is not in the document]
Comment (merrcury) [reply apps/worker/src/app/workflow/usecases/run-job/run-job.usecase.ts:140]: @greptileai Fixed in 3db131561e — the heartbeat is now created immediately before the try block, after the RUNNING step-run write, so nothing throwable sits between its creation and the finally that clears it. A failure in stepRunRepository.create now leaves no interval behind: the lease goes stale after 60s and redelivery reclaims the job. Covered by a regression test that fails on the previous ordering.
